src/mongocxx/pool.hpp

```diff
@@ -0,0 +1,128 @@
+/// to manually constructing @c client objects. The pool will use a single background thread used
+/// to monitor the topology of the replica set that is shared between all the client objects it
+/// creates. Each standalone client, however, will start its own background thread, leading to many
+/// redundant threads and network operations.
+///
+/// As a @c client is @b not thread safe, the @c pool should be preferred in multithreaded
+/// programs as it can safely be shared across threads.
+///
+class MONGOCXX_API pool {
+   public:
+    ///
+    /// Creates a pool associated with a connection string.
+    ///
+    /// @param mongodb_uri
+    ///  A MongoDB URI representing the connection parameters
+    /// @param ssl_options
+    ///  Optional SSL options to use when connecting to the MongoDB deployment.
+    pool(const uri& mongodb_uri = mongocxx::uri(),
+         stdx::optional<options::ssl> ssl_options = stdx::nullopt);
+
+    ///
+    /// Move constructs a pool.
+    ///
+    pool(pool&&) noexcept;
+
+    ///
+    /// Move assigns a pool.
+    ///
+    pool& operator=(pool&&) noexcept;
+
+    ///
+    /// Destroys a pool.
+    ///
+    ~pool();
+
+    ///
+    /// A unique_client is a handle on a @c client object acquired via the pool.
+    ///
+    /// @note The lifetime of any unique_client object must be a subset of the pool object
+    ///  from which it was acquired.
+    ///
+    using unique_client = std::unique_ptr<client, pool::release>;
+
+    ///
+    /// Acquires a client from the pool. The calling thread will block until a connection is
+    /// available.
+    ///
+    unique_client acquire();
+
+    ///
+    /// Acquires a client from the pool. This method will return immediately, but may return a
+    /// disengaged optional if a client is not available.
+    ///
+    stdx::optional<unique_client> try_acquire();
+
+    ///
+    /// Returns a client back to the pool.
+    ///
+    /// @warning This method must only be called on a client that was acquired from this pool.
+    ///   The result of calling release on a client that was constructed directly, or acquired from
+    ///   a different pool is undefined.
+    ///
+    /// @remark This method should not be called manually - the unique_client object returned from
+    ///   acquire will automatically call release when it is destroyed.
+    ///
+    void release(client* client);
+
+    ///
+    /// Gets a handle to the underlying implementation.
+    ///
+    /// Returned pointer is only valid for the lifetime of this object.
+    ///
+    /// @deprecated Future versions of the driver reserve the right to change the implementation
+    ///   and remove this interface entirely.
+    ///
+    /// @return Pointer to implementation of this object, or nullptr if not available.
+    ///
+    MONGOCXX_DEPRECATED void* implementation() const;
+
+   private:
+    class MONGOCXX_PRIVATE impl;
+    std::unique_ptr<impl> _impl;
```

Comment:
If you decide against move for this, just make this a const std::unique_ptr<impl> and that should auto inhibit move.